File: examples/python-fastapi/interaction_log.py
"""
Circular Buffer Interaction Log (Thread-Safe)

Memory-bounded log for tracking agent interactions.
Useful for debugging, analytics, and OASF heartbeat data.

Usage:
    from interaction_log import InteractionLog

    log = InteractionLog(max_size=1000)
    log.add(type="mcp", tool="getPrice", duration=150)
    recent = log.get_recent(10)
"""

import threading
from datetime import datetime, timezone
from typing import Any


class InteractionLog:
    def __init__(self, max_size: int = 1000):
        self._buffer: list[dict[str, Any] | None] = [None] * max_size
        self._max_size = max_size
        self._head = 0
        self._count = 0
        self._lock = threading.Lock()

    def add(self, **kwargs: Any) -> None:
        entry = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            **kwargs,
        }
        with self._lock:
            self._buffer[self._head] = entry
            self._head = (self._head + 1) % self._max_size
            if self._count < self._max_size:
                self._count += 1

    def get_recent(self, n: int = 10) -> list[dict[str, Any]]:
        with self._lock:
            result = []
            count = min(n, self._count)
            start = (self._head - count + self._max_size) % self._max_size
            for i in range(count):
                idx = (start + i) % self._max_size
                entry = self._buffer[idx]
                if entry is not None:
                    result.append(entry)
            return result

    def get_stats(self) -> dict[str, Any]:
        with self._lock:
            by_type: dict[str, int] = {}
            for i in range(self._count):
                idx = (self._head - self._count + i + self._max_size) % self._max_size
                entry = self._buffer[idx]
                if entry and "type" in entry:
                    t = entry["type"]
                    by_type[t] = by_type.get(t, 0) + 1
            return {"total": self._count, "by_type": by_type}

    def clear(self) -> None:
        with self._lock:
            self._head = 0
            self._count = 0

File: examples/typescript-hono/src/interaction-log.ts
/**
 * Circular Buffer Interaction Log
 *
 * Thread-safe, memory-bounded log for tracking agent interactions.
 * Useful for debugging, analytics, and OASF heartbeat data.
 *
 * Usage:
 *   const log = new InteractionLog(1000);
 *   log.add({ type: "mcp", tool: "getPrice", duration: 150 });
 *   const recent = log.getRecent(10);
 */

interface LogEntry {
  timestamp: string;
  type: string;
  [key: string]: unknown;
}

export class InteractionLog {
  private buffer: LogEntry[];
  private head = 0;
  private count = 0;

  constructor(private maxSize: number = 1000) {
    this.buffer = new Array(maxSize);
  }

  add(entry: Omit<LogEntry, "timestamp">): void {
    const full = {
      ...entry,
      timestamp: new Date().toISOString(),
    } as LogEntry;
    this.buffer[this.head] = full;
    this.head = (this.head + 1) % this.maxSize;
    if (this.count < this.maxSize) this.count++;
  }

  getRecent(n: number = 10): LogEntry[] {
    const result: LogEntry[] = [];
    const start = (this.head - Math.min(n, this.count) + this.maxSize) % this.maxSize;
    for (let i = 0; i < Math.min(n, this.count); i++) {
      const idx = (start + i) % this.maxSize;
      if (this.buffer[idx]) result.push(this.buffer[idx]);
    }
    return result;
  }

  getStats(): { total: number; byType: Record<string, number> } {
    const byType: Record<string, number> = {};
    for (let i = 0; i < this.count; i++) {
      const idx = (this.head - this.count + i + this.maxSize) % this.maxSize;
      const entry = this.buffer[idx];
      if (entry) {
        byType[entry.type] = (byType[entry.type] || 0) + 1;
      }
    }
    return { total: this.count, byType };
  }

  clear(): void {
    this.head = 0;
    this.count = 0;
  }
}
